Add some JFR event tests

roberttoyonaga opened the pull request:
Add some tests for recently added JFR events:
- jdk.JavaMonitorEnter
- jdk.JavaMonitorWait
- jdk.ThreadSleep


Each test generates the desired events to test, creates a JFR snapshot, and inspects the snapshot. There are 4 jdk.JavaMonitorWait tests that each cover a different case (time out, interruption, notify all, notify).

christianhaeubl commented on substratevm/src/com.oracle.svm.test/src/com/oracle/svm/test/jfr/TestJavaMonitorWaitTimeout.java:
```diff
@@ -0,0 +1,177 @@
+        notifierName = unheardNotifierThread.getName();
+
+        timeoutThread.start();
+        Thread.sleep(10);
```
This doesn't work reliably.

christianhaeubl commented on substratevm/src/com.oracle.svm.test/src/com/oracle/svm/test/jfr/TestThreadSleep.java:
```diff
@@ -0,0 +1,81 @@
+            if (!eventThread.equals(sleepingThreadName)) {
+                continue;
+            }
+            if (!isEqualDuration(event.getDuration(), Duration.ofMillis(MILLIS))) {
```
This won't work reliably.

christianhaeubl reviewed (changes requested):
Thanks for working on test cases for certain JFR events.

In a lot of your test cases, you try to use `Thread.sleep()` to ensure that multiple threads do something in a particular order. An approach like does not work reliably (especially if the machine has high load) and is therefore neither feasible for production code nor for test cases. You will need to switch to some kind of proper synchronization (e.g., `ReentrantLock` or a spin lock in case that you want to avoid that certain JFR events are emitted).

roberttoyonaga commented on substratevm/src/com.oracle.svm.test/src/com/oracle/svm/test/jfr/TestJavaMonitorEnter.java:
```diff
@@ -0,0 +1,112 @@
+        for (RecordedEvent event : events) {
+            RecordedObject struct = event;
+            String eventThread = struct.<RecordedThread> getValue("eventThread").getJavaName();
+            if (event.getEventType().getName().equals("jdk.JavaMonitorEnter") && isGreaterDuration(Duration.ofMillis(MILLIS), event.getDuration()) && waiterName.equals(eventThread)) {
```
I agree. I have changed this

roberttoyonaga commented on substratevm/src/com.oracle.svm.test/src/com/oracle/svm/test/jfr/JfrTest.java:
```diff
@@ -101,6 +111,9 @@ protected void enableEvents(String[] events) {
 
     public abstract String[] getTestedEvents();
 
+    public void analyzeEvents() {
```
changed to `validateEvents`

roberttoyonaga commented on substratevm/src/com.oracle.svm.test/src/com/oracle/svm/test/jfr/TestJavaMonitorEnter.java:
```diff
@@ -0,0 +1,112 @@
+                prev = duration;
+            }
+        }
+        assertTrue("Wrong number of Java Monitor Enter Events " + count, count == threads - 1);
```
Changed so that it filters based on the monitor and the threads are synchronized properly. The thread that is supposed to block, now should reliably wait for the other thread to enter the critical section.

roberttoyonaga commented on substratevm/src/com.oracle.svm.test/src/com/oracle/svm/test/jfr/TestJavaMonitorEnter.java:
```diff
@@ -0,0 +1,112 @@
+        try {
+            Stressor.execute(threadCount, eh, r);
+            // sleep so we know the event is recorded
+            Thread.sleep(500);
```
It's not needed. Removed now.

roberttoyonaga commented on substratevm/src/com.oracle.svm.test/src/com/oracle/svm/test/jfr/JfrTest.java:
```diff
@@ -127,6 +140,44 @@ protected void checkRecording() throws AssertionError {
+        return p;
+    }
+
+    protected List<RecordedEvent> getEvents(String testName) throws IOException {
```
I changed that method so that it will filter out all events except the ones that are on the list of tested events. This should allow each test to skip that step.

roberttoyonaga commented:
Thank you for the review Christian! I have updated the thread synchronization to be more reliable and removed the cases where sleeping is used to try to enforce ordering. Please let me know if there are still issues.

roberttoyonaga commented on substratevm/src/com.oracle.svm.test/src/com/oracle/svm/test/jfr/JfrTest.java:
```diff
@@ -101,6 +111,9 @@ protected void enableEvents(String[] events) {
+    public void analyzeEvents() {
+    }
+
     protected void checkEvents() {
```
I'm not sure why it's not being called. I add a call to it after the tests.

christianhaeubl reviewed (changes requested):
Thanks for the changes this already improves the situation. However, there is still one problematic pattern: querying `Thread.getState()` in a loop is unreliable because it doesn't tell the reason why a thread is blocked or waiting. So, the following can happen:
- Thread A calls `Thread.getState()` on thread B and checks if thread B is blocked or waiting.
- Thread B blocks or waits in JDK-internal code while the thread is being started (i.e. before it even reaches `run()`).
- Thread A sees the changed thread status and assumes that thread B is blocked at the right position in `run()`.

roberttoyonaga commented on substratevm/src/com.oracle.svm.test/src/com/oracle/svm/test/jfr/TestJavaMonitorEnter.java:
```diff
@@ -0,0 +1,121 @@
+                    Thread.sleep(10);
+                }
+                blockedAtCritical = true;
+                helper.doWork();
```
Is setting/checking of `blockedAtCritical` enough to guarantee the thread is blocking where we expect it to?

roberttoyonaga commented on substratevm/src/com.oracle.svm.test/src/com/oracle/svm/test/jfr/TestJavaMonitorEnter.java:
```diff
@@ -0,0 +1,121 @@
+        Runnable second = () -> {
+            try {
+                // wait until lock is held
+                while (!inCritical) {
```
That's a good idea. Done!

christianhaeubl reviewed (approved):
Thanks, your PR will be merged to master in the next couple days.
